syncer/: fix flush checkpoint for ddls for situations that IsSharding is false

lichunzhu opened the pull request:
<!--
Thank you for contributing to DM! Please read MD's [CONTRIBUTING](https://github.com/pingcap/dm/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->
https://github.com/pingcap/dm/issues/514
* When `IsSharding` is false, table checkpoint which contains schema tracker info was not updated in `dm_meta` instantly. But global checkpoint is updated. If worker fails now, it will cause a mismatch between tracker and global checkpoint.
* Fix a dead lock bug.

### What is changed and how it works?
Update table point when we update global checkpoint.

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

lichunzhu commented:
Should be merged after https://github.com/pingcap/dm/pull/581/files, add uncomment the blocked integration tests

csuzhangxc commented on syncer/syncer.go:
```diff
@@ -742,15 +742,19 @@ func (s *Syncer) addJob(job *job) error {
-			s.saveTablePoint(job.sourceSchema, job.sourceTable, job.currentLocation)
+		for i := range job.sourceSchemas {
+			if len(job.sourceSchemas[i]) > 0 {
+				s.saveTablePoint(job.sourceSchemas[i], job.sourceTables[i], job.location)
```
Should we use `job. location` or `job.currentLocation` here?

lichunzhu replied:
updated in https://github.com/pingcap/dm/pull/587/commits/f1837f8d8dd5d9aba2c9442fc5f5f9326745a019

csuzhangxc commented:
in #514, I think we should also **flush** (not only **save**) checkpoint **before** executing DDL to solve a case:

1. `INSET TABLE tbl (c1) VALUES (1)`
2. `ALTER TABLE tbl ADD COLUMN (c2 INT)`
3. flush checkpoint into downstream TiDB

interrupted after step.2 and before step.3, and then restarted from step.1, this will trigger an error with `columns and data mismatch in length: 2 (columns) 1 (data)`

lichunzhu commented:
> in #514, I think we should also **flush** (not only **save**) checkpoint **before** executing (and track) DDL to solve a case:
> 
> 1. `INSET TABLE tbl (c1) VALUES (1)`
> 2. `ALTER TABLE tbl ADD COLUMN (c2 INT)`
> 3. flush checkpoint into downstream TiDB
> 
> interrupted after step.2 and before step.3, and then restarted from step.1, this will trigger an error with `columns and data mismatch in length: 2 (columns) 1 (data)`

updated in https://github.com/pingcap/dm/pull/587/commits/f1837f8d8dd5d9aba2c9442fc5f5f9326745a019. A test is also added in this commit.

csuzhangxc commented on syncer/optimist.go:
```diff
@@ -78,6 +78,10 @@ func (s *Syncer) handleQueryEventOptimistic(
 		}
 	}
 
+	if err := s.flushCheckPoints(); err != nil {
```
Can we put it near `s.jobWg.Wait()` to make it more meaningful?

csuzhangxc commented on syncer/syncer.go:
```diff
@@ -753,11 +753,16 @@ func (s *Syncer) addJob(job *job) error {
 		}
 	}
 
+	failpoint.Inject("FlushCheckpointExit", func() {
```
in many places, we call `s.flushCheckPoints` directly, but they will not trigger this `FlushCheckpointExit`, how about giving it another name or move it into `s.flushCheckPoints`?

csuzhangxc commented on syncer/syncer.go:
```diff
@@ -1706,6 +1711,10 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext) e
 			needHandleDDLs = appliedSQLs // maybe nil
 		}
 
+		if err := s.flushCheckPoints(); err != nil {
```
If we do not dispatch a `flush` job, then we must call `s.jobWg.Wait()` before `s.flushCheckPoints()`.

(Is dispatch a `flush` job better?)

csuzhangxc commented on tests/sharding2/run.sh:
```diff
@@ -46,6 +46,40 @@ function run() {
+    echo "kill dm-worker1"
+    ps aux | grep dm-worker1 |awk '{print $2}'|xargs kill || true
+    echo "kill dm-worker2"
+    ps aux | grep dm-worker2 |awk '{print $2}'|xargs kill || true
```
Do we need to `check_port_offline` here?

csuzhangxc commented:
/run-all-tests

csuzhangxc commented on syncer/syncer.go:
```diff
@@ -1799,7 +1798,7 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext) e
 	s.tctx.L().Info(annotate, zap.String("event", "query"), zap.String("source", source), zap.Strings("ddls", needHandleDDLs), zap.ByteString("raw statement", ev.Query), zap.Bool("in-sharding", needShardingHandle), zap.Stringer("start location", startLocation), zap.Bool("is-synced", synced), zap.Int("unsynced", remain))
 
-	if err := s.flushCheckPoints(); err != nil {
+	if err := s.flushJobs(); err != nil {
```
It seems we can remove `s.jobWg.Wait()` in L1856 now? (or even we can remove `s.jobWg.Wait()` in L723? but we may still just keep it, the shard mechanism make this not uniform, 😢 )

WangXiangUSTC commented on syncer/job.go:
```diff
@@ -58,8 +58,8 @@ func (t opType) String() string {
-	sourceSchema    string
-	sourceTable     string
+	sourceSchemas   []string
+	sourceTables    []string
```
why job need to use multiple schemas and tables? can you add a comment

WangXiangUSTC replied:
and I think map is better

lichunzhu replied:
all updated in https://github.com/pingcap/dm/pull/587/commits/2733185c2787894f9104e20aab935ac154092e4b

WangXiangUSTC commented on syncer/job.go:
```diff
@@ -111,10 +112,21 @@ func newDDLJob(ddlInfo *shardingDDLInfo, ddls []string, location, cmdLocation bi
+		j.sourceTables = []string{ddlInfo.tableNames[0][0].Name}
 		j.targetSchema = ddlInfo.tableNames[1][0].Schema
 		j.targetTable = ddlInfo.tableNames[1][0].Name
+	} else if sourceTbls != nil {
```
in which case `ddlInfo` is nil and `sourceTbls` is not nil? can you add a example

lichunzhu replied:
updated in https://github.com/pingcap/dm/pull/587/commits/2733185c2787894f9104e20aab935ac154092e4b
